OpenRA.Mods.Common/UtilityCommands/ExtractEmmyLuaAPI.cs

```diff
@@ -0,0 +1,358 @@
+			Console.WriteLine("---@field X integer");
+			Console.WriteLine("---@field Y integer");
+			Console.WriteLine("---@field Z integer");
+			Console.WriteLine();
+			Console.WriteLine("---@class cvec");
+			Console.WriteLine("---@field X integer");
+			Console.WriteLine("---@field Y integer");
+			Console.WriteLine();
+			Console.WriteLine("---@class color");
+			Console.WriteLine("local color = { };");
+		}
+
+		static void WriteActorInits(IEnumerable<Type> actorInits, out IEnumerable<Type> usedEnums)
+		{
+			var inits = actorInits.Select(x =>
+				(x.Name.Substring(0, x.Name.Length - 4), x.GetConstructors()));
+
+			Console.WriteLine("---A list of ActorInit implementations that can be used by Lua scripts.");
+			Console.WriteLine("---@class initTable");
+
+			var localEnums = new List<Type>();
+			foreach (var (name, constructorInfos) in inits)
+			{
+				var parameters = constructorInfos.Select(ci => ci.GetParameters());
+				var parameterString = string.Join(" | ",
+					parameters
+						.Select(cp => string.Join(", ",
+							cp
+								.Where(p => !p.HasDefaultValue)
+								.Where(p => p.ParameterType != typeof(TraitInfo) && p.ParameterType.Name != typeof(Func<int>).Name)
+								.Select(p =>
+								{
+									if (p.ParameterType.IsEnum)
+										localEnums.Add(p.ParameterType);
+
+									return p.ParameterType.EmmyLuaString();
+								})))
+						.Where(s => !s.Contains(", "))
+						.Distinct());
+
+				if (!string.IsNullOrEmpty(parameterString))
+					Console.WriteLine($"---@field {name} {parameterString}");
+			}
+
+			usedEnums = localEnums.Distinct();
+		}
+
+		static void WriteEnums(IEnumerable<Type> enumTypes)
+		{
+			foreach (var enumType in enumTypes)
+			{
+				Console.WriteLine(enumType.Name + " = {");
+
+				foreach (var value in Enum.GetValues(enumType))
+					Console.WriteLine($"    {value} = {Convert.ChangeType(value, typeof(int))},");
+
+				Console.WriteLine("}");
+				Console.WriteLine();
+			}
+		}
+
+		static void WriteGlobals(IEnumerable<Type> globalTables)
+		{
+			foreach (var t in globalTables)
+			{
+				var name = t.GetCustomAttributes<ScriptGlobalAttribute>(true).First().Name;
+				Console.WriteLine("---Global variable provided by the game scripting engine.");
+
+				foreach (var obsolete in t.GetCustomAttributes(false).OfType<ObsoleteAttribute>())
+				{
+					Console.WriteLine("---@deprecated");
+					Console.WriteLine($"--- {obsolete.Message}");
+				}
+
+				Console.WriteLine(name + " = {");
+
+				var members = ScriptMemberWrapper.WrappableMembers(t);
+				foreach (var member in members.OrderBy(m => m.Name))
+				{
+					Console.WriteLine();
+
+					var body = "";
+
+					if (member.HasAttribute<DescAttribute>())
+					{
+						var lines = member.GetCustomAttributes<DescAttribute>(true).First().Lines;
+						foreach (var line in lines)
+							Console.WriteLine($"    --- {line}");
+					}
+
+					var propertyInfo = member as PropertyInfo;
+					if (propertyInfo != null)
```

Comment:
C# 7 pattern matching is now used as of PR #20020:
```suggestion
					if (member is PropertyInfo propertyInfo)
```